plugins/user.py

```diff
@@ -0,0 +1,507 @@
+        else:
+            self.direct_reply(message, "User '{}' has permissions {}".format(user.nick, ', '.join(user.permissions)))
+
+    @respond_to("^who can (?P<permission>([\w.:-]+))")
+    @warn_direct_message
+    @requires_permission('view_permissions')
+    def get_users_for_permissions(self, message, permission):
+        """who can [permission]?: see which users have a particular permission (requires the 'view_permissions' permission)"""
+
+        nicks_with_perm = []
+        for user in User.list(self):
+            if user.has_permission(permission):
+                nicks_with_perm.append(user.nick)
+
+        if len(nicks_with_perm) > 0:
+            self.direct_reply(message, "The following users have permission to '{}': {}".format(permission, ', '.join(nicks_with_perm)))
+        else:
+            self.direct_reply(message, "No users have permission to '{}'".format(permission))
+
+    @respond_to("^can I(?P<permissions>( [\w.:-]+)+)")
+    @warn_direct_message
+    def confirm_user_permission(self, message, permissions):
+        """can I [permission]?: check if you have a specific permission"""
+        user = User.get_from_message(self, message)
+        if not user:
+            return
+
+        if not user.is_authenticated:
+            self.direct_reply(message, "I am unable to check your permissions because you are not authenticated. Please start an authenticated session by saying 'twofactor verify [token]'.")
+            return
+
+        for permission in permissions.split():
+            if user.has_permission(permission):
+                self.direct_reply(message, "Yes, you can {}".format(permission))
+            else:
+                self.direct_reply(message, "No, you can't {}".format(permission))
+
+    @respond_to("^give (?P<nick>\w+) permission(?P<permissions>( [\w.:-]+)+)")
+    @warn_direct_message
+    @requires_permission('grant_permissions')
+    def give_user_permission(self, message, nick, permissions):
+        """give [nick] permission to [permission]: grant [nick] a permission (requires the 'grant_permissions' permission)"""
+        requested_permissions = permissions.split()
+        try:
+            requested_permissions.remove("to")
+        except ValueError:
+            pass
+
+        if len(requested_permissions) == 0:
+            self.direct_reply(message, 'At least one permission must be specified')
+            return
+
+        user = User.get_from_nick(self, nick)
+        if not user:
+            self.direct_reply("The user '{}' has not setup two-factor authentication, please have them do so before modifying permissions")
+            return
+
+        user.grant_permissions(requested_permissions)
+        user.save()
+
+        self.direct_reply(message, "New permissions for '{}' are: {} ".format(nick, ', '.join(user.permissions)))
+
+    @respond_to("^take away from (?P<nick>\w+) permission(?P<permissions>( [\w.:-]+)+)")
```

Comment:
I had trouble getting alton to parse "text (argument)+ text (argument)"  - pulling out n+1 words from the middle of a string is harder than pulling them off of the end.